### PR TITLE
Fixed DP-2124. Implemented SYNCHRONIZE CACHE cmd

### PR DESCRIPTION
**Pull request by dmehta85**
Please refer JIRA ticket to understand SYNC cmd and its implmentation
in detail at thrift-client side using,
1. SYNC command Design Document
2. SCSI command Reference Manual

Signed-off-by: Deepen Mehta <deepen.mehta@primaryio.com>

### CODE REVIEW

File: src/include/Request.h [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+	Request(RequestID id, Type t, const void* privatep, char *bufferp,
+		int32_t buf_sz, uint64_t length, int64_t offset);
+
+	~Request();
+};
+
+class SyncRequest : public RequestBase {
+public:
+	uint32_t count;
+	std::vector<std::weak_ptr<RequestBase>> write_pending;
```
Comment (prasad-joshi) [inline src/include/Request.h]: Why are these weak_ptr<>. These should be shared_ptrs

File: src/thrift-client/TgtInterfaceImpl.cpp [hunk 2]
```diff
@@ -781,31 +737,71 @@ int64_t StordVmdk::RpcRequestScheduledCount() const noexcept {
...
-		latency_avg_.Average() > kExpectedWanLatency;
+
+	/* Effective for new writes overlapping on sync */
+	bool overlapped_write = false;
+	if (nreqp->type == Request::Type::kWrite or
+		nreqp->type == Request::Type::kWriteSame) {
+		for (auto& sync_req : requests_.sync_pending_) {
+			SyncRequest *syncp =
+				dynamic_cast<SyncRequest *>(sync_req.second.get());
+			if (nreqp->IsOverlapped(syncp->offset, syncp->length)) {
```
Comment (prasad-joshi) [inline src/thrift-client/TgtInterfaceImpl.cpp]: Incorrect use of dynamic_cast. It is not required here.

File: src/thrift-client/TgtInterfaceImpl.cpp [hunk 3]
```diff
@@ -848,6 +844,87 @@ void StordVmdk::UpdateStats(Request* reqp) {
...
+}
+
+SyncRequest::SyncRequest(RequestID id, Type t, const void* privatep,
+	uint64_t length, int64_t offset, StordVmdk* vmdkp) :
+	RequestBase(id, t, privatep, length, offset), vmdkp(vmdkp) {
+}
+
+SyncRequest::~SyncRequest() {
+	for (auto write_req : write_pending) {
+		Request *reqp = reinterpret_cast<Request *>(write_req.lock().get());
```
Comment (prasad-joshi) [inline src/thrift-client/TgtInterfaceImpl.cpp]: weak_ptr.lock() does not guarantee returned values is shared_ptr.

File: src/thrift-client/TgtInterfaceImpl.cpp [hunk 4]
```diff
@@ -781,31 +737,71 @@ int64_t StordVmdk::RpcRequestScheduledCount() const noexcept {
...
+	bool prepared = true;
+	auto nreqp = request.get();
+
+	if (nreqp->type != Request::Type::kSync) {
+		return prepared;
 	}
 
-	++stats_.pending_;
+	std::lock_guard<std::mutex> lock(requests_.mutex_);
+	SyncRequest *reqp = dynamic_cast<SyncRequest*>(nreqp);
```
[the remark this answers is not in the document]
Comment (dmehta85) [reply src/thrift-client/TgtInterfaceImpl.cpp]: Yes. Will remove repeated used.

File: src/thrift-client/TgtInterfaceImpl.cpp [hunk 5]
```diff
@@ -848,6 +844,87 @@ void StordVmdk::UpdateStats(Request* reqp) {
...
+RequestBase::~RequestBase() {
+}
+
+const RequestBase::Type& RequestBase::GetType() const noexcept {
+		return type;
+}
+
+bool RequestBase::IsOverlapped(uint64_t req_offset,
+	uint64_t req_length) const noexcept {
+	return ((length - req_offset > 0) && (req_length - offset > 0));
```
[the remark this answers is not in the document]
Comment (dmehta85) [reply src/thrift-client/TgtInterfaceImpl.cpp]: True. Will take subtraction value in int64_t  variable.

File: src/thrift-client/TgtInterfaceImpl.cpp (hunk 2 above)
Comment (dmehta85) [reply src/thrift-client/TgtInterfaceImpl.cpp]: Will remove it.

File: src/thrift-client/TgtInterfaceImpl.cpp (hunk 3 above)
Comment (dmehta85) [reply src/thrift-client/TgtInterfaceImpl.cpp]: Will update to shared ptr. so, locking will not be required.

File: src/include/Request.h (hunk 1 above)
Comment (dmehta85) [reply src/include/Request.h]: will update.

**Comment by dmehta85** [comment]
Incorporated all review comments given by Prasad.

File: src/thrift-client/TgtInterfaceImpl.cpp [hunk 6]
```diff
@@ -848,6 +844,88 @@ void StordVmdk::UpdateStats(Request* reqp) {
...
+}
+
+const RequestBase::Type& RequestBase::GetType() const noexcept {
+		return type;
+}
+
+bool RequestBase::IsOverlapped(uint64_t req_offset,
+	uint64_t req_length) const noexcept {
+	return ((int64_t)(length - req_offset) > 0 &&
+		(int64_t)(req_length - offset) > 0);
```
Comment (prasad-joshi) [inline src/thrift-client/TgtInterfaceImpl.cpp]: @dmehta85 this condition is wrong. Please test this code in isolation.

Comment (prasad-joshi) [reply src/thrift-client/TgtInterfaceImpl.cpp]: Typecasting uint64_t to int64_t should also be avoided.

Comment (dmehta85) [reply src/thrift-client/TgtInterfaceImpl.cpp]: Based on the section 'How to check writes are overlapped' given in the [design document](https://primaryio.atlassian.net/secure/attachment/12388/SyncDesign.pdf), I have written an isolated program and found a condition works perfectly. 
[Isolated Program](https://primaryio.atlassian.net/secure/attachment/12397/check_overlap.c)  uploaded on jira ticket [DP-2124](https://primaryio.atlassian.net/browse/DP-2124)

Program output:
#> ./check_overlap 
Write[1] is overlapped: No
Write[2] is overlapped: Yes
Write[3] is overlapped: Yes
Write[4] is overlapped: Yes
Write[5] is overlapped: Yes
Write[6] is overlapped: No
Write[7] is overlapped: Yes
#>

Please share the condition in which above logic becomes false.

Comment (prasad-joshi) [reply src/thrift-client/TgtInterfaceImpl.cpp]: 1. Assume, existing request  has Offset = 1M (1048576), Length = 4096
2. New request is req_offset = 1M + 512 = 1049088 and Length = 512

These are overlapping requests 

1. length - req_offset > 0 = 4096 - (1M+512)  > 0 = false
2. req_length - offset > 0 = 512 - (1M) > 0 = false

Your program checks with pretty small numbers.

Comment (dmehta85) [reply src/thrift-client/TgtInterfaceImpl.cpp]: Calculation is corrected.

**Comment by dmehta85** [comment]
> We observed a deadlock yesterday. I think, code needs change.

Resolved. All details are mentioned in commit description. 
Did the following testing:
1. Migrated Windows-2016 VM
2. Booted successfully on-cloud. Performed some IO(s)
3. Migrated-back Windows-2016 VM
4. Booted successfully on-prem. Verify IO(s)

**Pull request merged**